### PR TITLE
Assignment9+10 @hgill17

### PR DESCRIPTION
**Pull request by dcox2005**
Assignment9+10

### CODE REVIEW

**Comment by jmplett** [comment]
I will review

**Comment by kelseymcmahon** [comment]
## Instructions

1. Implement `PingProcess`' `public Task<PingResult> RunTaskAsync(string hostNameOrAddress)` ✔
   - First implement `public void RunTaskAsync_Success()` test method to test `PingProcess.RunTaskAsync()` using `"localhost"`. ✔
   - Do NOT use async/await in this implementation. ✔
2. Implement `PingProcess`' `async public Task<PingResult> RunAsync(string hostNameOrAddress)` ✔
   - First implement the `public void RunAsync_UsingTaskReturn_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` **without** using async/await. ✔
   - Also implement the `async public Task RunAsync_UsingTpl_Success()` test method to test `PingProcess.RunAsync()` using `"localhost"` but this time **DO** using async/await. ✔
3. Add support for an optional cancellation token to the `PingProcess.RunAsync()` signature. ✔
   Inside the `PingProcess.RunAsync()` invoke the token's `ThrowIfCancellationRequested()` method so an exception is thrown. ✔
   Test that, when cancelled from the test method, the exception thrown is an `AggregateException` ✔ with a `TaskCanceledException` inner exception ✔ using the test methods `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrapping` ✔and `RunAsync_UsingTplWithCancellation_CatchAggregateExceptionWrappingTaskCanceledException` ✔ respectively.
4. Complete/fix **AND test** `async public Task<PingResult> RunAsync(IEnumerable<string> hostNameOrAddresses, CancellationToken cancellationToken = default)` which executes ping for and array of hostNameOrAddresses (which can all be "localhost") **in parallel**, adding synchronization if needed. ✔
   NOTE:
      - The order of the items in the stdOutput is irrelevent and expected to be intermingled.
      - StdOutput must have all the ping output returned (no lines can be missing) even though intermingled. ✔
5. Implement **AND test** `public Task<int> RunLongRunningAsync(ProcessStartInfo startInfo, Action<string?>? progressOutput, Action<string?>? progressError, CancellationToken token)` using `Task.Factory.StartNew()` and invoking `RunProcessInternal` with a `TaskCreation` value of `TaskCreationOptions.LongRunning` and a `TaskScheduler` value of `TaskScheduler.Current`.
   NOTE: This method does **NOT** use `Task.Run`. ✔

## Extra Credit

- Test and implement `PingProcess.RunAsync(System.IProgress<T> progess)` so that you can capture the output as it occurs rather than capturing the output only when the process finishes. ❌

## Fundamentals

- Ensure you enable:
  - nullable reference types is enabled ❌
  - net6 targeted ❌
  - C# 10.0 ❌
  - and enabled .NET analyzers for both projects ❌
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

Overall everything looked great! I liked how your group used the ExceptionDispatchInfo to throw the exception for the test to catch, we could not figure that out and did it with Assert. 

I would comment that in the RunAsync_UsingTaskReturn_Success() test method a cancellation token is created, sent, and then cancelled. I don't think it is necessary to cancel the task here and this could be dropped. 

Also, don't forget to edit the project files to include the fundamentals that are listed above.

File: Assignment/Assignment.Tests/PingProcessTests.cs [hunk 1]
```diff
@@ -3,6 +3,8 @@
 using System.Collections.Generic;
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/PingProcessTests.cs:3]: You should have nullability and analyzers turned on

File: Assignment/Assignment/PingProcess.cs [hunk 2]
```diff
@@ -27,40 +27,46 @@ public PingResult Run(string hostNameOrAddress)
...
 
     async public Task<PingResult> RunLongRunningAsync(
         string hostNameOrAddress, CancellationToken cancellationToken = default)
     {
-        Task task = null!;
+        Task<PingResult> task = Task.Factory.StartNew(
+            () => Run(hostNameOrAddress), cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
         await task;
-        throw new NotImplementedException();
+        return task.Result;
```
Comment (mmwoodfo) [inline Assignment/Assignment/PingProcess.cs:69]: You can just return the task here and remove the `async` from your method

**Pull request closed without merging**